mne/tests/test_epochs.py

```diff
@@ -2450,6 +2450,44 @@ def test_events_list():
     assert_array_equal(epochs.events, np.array(events))
 
 
+def test_save_overwrite(tmpdir):
+    """Test saving with overwrite functionality."""
+
+    tempdir = str(tmpdir)
+    raw = mne.io.RawArray(np.random.RandomState(0).randn(100, 10000),
+                          mne.create_info(100, 1000.))
+
+    events = mne.make_fixed_length_events(raw, 1)
+    epochs = mne.Epochs(raw, events)
+
+    # scenario 1: overwrite=False (default) and there isn't a file to overwrite
+    # make a filename that has not already been saved to
+    fname1 = op.join(tempdir, 'test_v1-epo.fif')
+    # run function to be sure it doesn't throw an error
+    epochs.save(fname1)
+    # check that the file got written
+    assert(op.isfile(fname1))
+
+    # scenario 2: overwrite=False (default) and there is a file to overwrite
+    # fname1 exists because of scenario 1 above
+    pytest.raises(IOError, epochs.save, fname1)
```

Comment:
Via proper deprecation this shouldn't raise an error currently but rather a warning. So this should be:
```
with pytest.deprecated_call():
    epochs.save(fname1)
```
And then `epochs.save(fname1, overwrite=False)` should raise an error, but let's throw (and catch) `FileExistsError` so that you can write:
```
with pytest.raises(FileExistsError, match='some matching substring):
    epochs.save(fname1)
```
... and modify `_check_fname` to throw `FileExistsError` instead of plan `OSError`.

Reply:
I edited the tests - thanks for the sample code! I think that adding `FileExistsError` to `_check_fname` is out of scope for this PR. I'm happy to open a new issue for that if you like, and provided @agramfort agrees?

Reply:
Yes we can do it separately. We might want one big PR to modernize our error messages with the new subclasses